proof/refine/ARM/StateRelation.thy

```diff
@@ -880,33 +891,66 @@ lemma sc_replies_relation_prev_list:
   apply (clarsimp simp: sc_of_def opt_map_def projectKO_opt_sc split: option.splits)
   done
 
-(* FIXME: can generalise in terms of "sym heaps", maybe later *)
-lemma sym_refs_prev_eq_next_sc_replies:
-  "\<lbrakk> sym_refs (list_refs_of_replies' s');
-     heap_list (replyPrevs_of s') (scReply sc') replies\<rbrakk>
-    \<Longrightarrow> heap_list (replyNexts_of s') (scReply sc') (rev replies)"
-  oops
+lemma list_refs_of_replies'_reftype[simp]:
+  "(p, reftype) \<in> list_refs_of_replies' s' p' \<Longrightarrow> reftype \<in> {ReplyPrev, ReplyNext}"
+  by (clarsimp simp: list_refs_of_replies'_def list_refs_of_reply'_def get_refs_def2
+              split: option.split_asm)
+
+(* FIXME RT: move *)
+lemma opt_map_SomeD:
+  "f x = Some y \<Longrightarrow> (f |> g) x = g y"
+  by (clarsimp simp: opt_map_Some_def)
+
+lemma replyNext_replyNexts_of_opt_map[simp]:
+  "\<lbrakk>ksPSpace s' p = Some (KOReply reply); replyNext reply = Some (Next p')\<rbrakk>
+    \<Longrightarrow> (replyNexts_of s' |> replyPrevs_of s') p = replyPrevs_of s' p'"
```

Comment:
Could both instances of `replyPrevs_of s'` in this theorem be replaced with some fully generalised function `f`?

Reply:
yes, that’s right. This is all about `replyNext` and `Next`. I’ll generalise the second function.